build-system/babel-plugins/babel-plugin-transform-log-methods/index.js

```diff
@@ -0,0 +1,320 @@
+ * of a log message is ~43 whereas the common minified transformed output (as
+ * `x().l('a0')`) is just 11. Since the prefix of the output could repeat
+ * throughout a binary, it also compresses well.
+ *
+ * Resulting compressed, minified binaries reduce size by ~1.5% depending on
+ * their logging density.
+ */
+const fs = require('fs');
+const path = require('path');
+const {
+  messagesPath,
+  singletonFunctions,
+  transformableMethods,
+} = require('../../log-module-metadata.js');
+
+/**
+ * Path to messages file for this version relative to the repo root.
+ * This is a default since tests can configure this table's output path, but for
+ * all building purposes this should be considered the canonical messages
+ * filepath for the built version.
+ */
+const defaultMessagesPath = messagesPath;
+
+/** Method on `Log` that this transform outputs to "expand" log messages. */
+const messageExpansionMethod = 'expandLogMessage';
+
+/**
+ * Mangled expansion method name to prevent large public method name in
+ * multipass build.
+ */
+const messageExpansionMethodShortName = 'l';
+
+/**
+ * Approximate length of a compressed message expansion call to determine
+ * whether an instance of message indirection actually reduces binary size.
+ */
+const roughMinifiedExpansionLength = 'a().b("xx")'.length;
+
+/**
+ * Determines a filepath relative to the repo root.
+ * @param {string} path
+ * @return {string}
+ */
+const relativeToRoot = path => `${__dirname}/../../../${path}`;
+
+/**
+ * Reads the messages table from the JSON file.
+ * @param {string} messagesPath
+ * @return {!Object<string, string>}
+ */
+function getMessages(messagesPath) {
+  try {
+    return JSON.parse(fs.readFileSync(messagesPath));
+  } catch {
+    // When non-existent or empty just return an empty object. Transformations
+    // that read will write as well.
+    return {};
+  }
+}
+
+/**
+ * Writes the messages table to the JSON file.
+ * @param {string} messagesPath
+ * @param {!Object<string, string>} obj
+ */
+function writeMessages(messagesPath, obj) {
+  const json = JSON.stringify(obj, /* replacer */ null, /* spaces */ 2);
+  const messagesPathDir = path.dirname(messagesPath);
+  if (!fs.existsSync(messagesPathDir)) {
+    fs.mkdirSync(messagesPathDir);
+  }
+  fs.writeFileSync(messagesPath, json);
+}
+
+/**
+ * Gets a message id from the table, or adds a new entry for a message if
+ * non-existent and returns its new id.
+ * @param {string} messagesPath
+ * @param {string} message
+ * @return {string} Short message id.
+ */
+function getOrCreateShortMessageId(messagesPath, message) {
+  const messages = getMessages(messagesPath);
+  if (messages[message]) {
+    return messages[message];
+  }
+  // Base-36 radix for best utilization of ascii space.
+  const shortMessageId = Object.keys(messages).length.toString(36);
+  messages[message] = shortMessageId;
+  writeMessages(messagesPath, messages);
```

Comment:
It's expensive to write a file every time we add a message. Try adding a `Program` visitor (with `enter`/`exit` methods) and write when we exit the `Program` (ie, when babel is done transforming an individual file).